
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Configuring the workflow &#8212; AmpSeeker-documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "sanjaynagi/ampseeker");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/configuring_the_workflow';</script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Running the workflow" href="running_the_workflow.html" />
    <link rel="prev" title="Input files" href="input_data.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../landing-page.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="AmpSeeker-documentation - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="AmpSeeker-documentation - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../landing-page.html">
                    AmpSeeker
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="input_data.html">Input files</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Configuring the workflow</a></li>



<li class="toctree-l1"><a class="reference internal" href="running_the_workflow.html">Running the workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakemake.html">Snakemake</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting and contributing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to the workflow</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/sanjaynagi/ampseeker" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sanjaynagi/ampseeker/issues/new?title=Issue%20on%20page%20%2Fnotebooks/configuring_the_workflow.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/configuring_the_workflow.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Configuring the workflow</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Configuring the workflow</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#configuration-parameters">Configuration Parameters</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cohort-analysis-configuration">Cohort Analysis Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-files-configuration">Input Files Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#illumina-directory-configuration">Illumina Directory Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-genome-configuration">Reference Genome Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-file-type-configuration">Input File Type Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-control-configuration">Quality Control Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis-configuration">Analysis Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#jupyter-book-configuration">Jupyter Book Configuration</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#special-analyses-for-ag-vampir-panel">Special Analyses for Ag-vampIR Panel</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example-configuration">Example Configuration</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#illumina-configuration-example">Illumina Configuration Example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#nanopore-configuration-example">Nanopore Configuration Example</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="configuring-the-workflow">
<h1>Configuring the workflow<a class="headerlink" href="#configuring-the-workflow" title="Link to this heading">#</a></h1>
<p>Before running AmpSeeker, we need to select which analyses we want to run (configuration). This is done by editing the file <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> in the <code class="docutils literal notranslate"><span class="pre">config</span></code> directory. The config file contains a number of options and parameters.</p>
<p><strong>If you have any issues configuring the pipeline, please watch the video walkthrough first, and raise an issue on github or <a class="reference external" href="mailto:sanjay&#46;c&#46;nagi&#37;&#52;&#48;gmail&#46;com">email me</a>.</strong></p>
</section>
<section id="configuration-parameters">
<h1>Configuration Parameters<a class="headerlink" href="#configuration-parameters" title="Link to this heading">#</a></h1>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">lab-strains</span>
<span class="nt">panel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ag-vampir</span>
<span class="nt">platform</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">illumina</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>dataset</strong>: A name for your dataset. This will be used to name output files and directories.</p></li>
<li><p><strong>panel</strong>: The name of the amplicon panel used. Currently, the workflow has special support for the “Ag-vampIR” panel (Anopheles gambiae vector amplicon marker panel for Insecticide Resistance). If not using Ag-vampIR, panel name can be anything.</p></li>
<li><p><strong>platform</strong>: The sequencing platform used. Options are <code class="docutils literal notranslate"><span class="pre">illumina</span></code> for Illumina sequencing (default) or <code class="docutils literal notranslate"><span class="pre">nanopore</span></code> for Nanopore sequencing. This determines the alignment and variant calling pipeline used.</p></li>
</ul>
<section id="cohort-analysis-configuration">
<h2>Cohort Analysis Configuration<a class="headerlink" href="#cohort-analysis-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">cohort-columns</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">location</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">taxon</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>cohort-columns</strong>: List of metadata columns used to group samples for analyses. These columns will be used to color samples in plots and to perform comparisons between groups. Common options include location, taxon, country, and any other categorical variables in your metadata.</p></li>
</ul>
</section>
<section id="input-files-configuration">
<h2>Input Files Configuration<a class="headerlink" href="#input-files-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">targets</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/ag-vampir.bed</span>
<span class="nt">metadata</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/metadata.tsv</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>targets</strong>: Path to a BED file containing the genomic coordinates of amplicon targets. This file should have 5 columns: chromosome, start, end, amplicon_id, and target_label with no header. See the input_data.ipynb for more details.</p></li>
<li><p><strong>metadata</strong>: Path to a TSV file containing sample metadata. At minimum, this file should have a “sample_id” column. Additional columns can be used for cohort analysis. Only required if working directly from fastq files (<code class="docutils literal notranslate"><span class="pre">from-bcl:</span> <span class="pre">False</span></code>).</p></li>
</ul>
</section>
<section id="illumina-directory-configuration">
<h2>Illumina Directory Configuration<a class="headerlink" href="#illumina-directory-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">illumina-dir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/250110_M05658_0028_000000000-LTBV4</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>illumina-dir</strong>: Path to the Illumina MiSeq run directory containing BCL files. This is only required if converting from BCL to FASTQ (<code class="docutils literal notranslate"><span class="pre">from-bcl:</span> <span class="pre">True</span></code>). If you already have FASTQ files, this can be left empty.</p></li>
</ul>
</section>
<section id="reference-genome-configuration">
<h2>Reference Genome Configuration<a class="headerlink" href="#reference-genome-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">reference-fasta</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/reference/Anopheles-gambiae-PEST_CHROMOSOMES_AgamP4.fa</span>
<span class="nt">reference-gff3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/reference/Anopheles-gambiae-PEST_BASEFEATURES_AgamP4.12.gff3</span>
<span class="nt">reference-snpeffdb</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Anopheles_gambiae</span>
<span class="nt">custom-snpeffdb</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>reference-fasta</strong>: Path to the reference genome FASTA file (can be gzipped with .fa.gz extension).</p></li>
<li><p><strong>reference-gff3</strong>: Path to the genome annotation file in GFF3 format.</p></li>
<li><p><strong>reference-snpeffdb</strong>: Name of the SnpEff database to use for variant annotation. This should match a database name available in SnpEff (e.g., “Anopheles_gambiae”).</p></li>
<li><p><strong>custom-snpeffdb</strong>: Whether to build a custom SnpEff database. Set to <code class="docutils literal notranslate"><span class="pre">True</span></code> if the reference genome is not available in the standard SnpEff databases.</p></li>
</ul>
</section>
<section id="input-file-type-configuration">
<h2>Input File Type Configuration<a class="headerlink" href="#input-file-type-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">from-bcl</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="nt">fastq</span><span class="p">:</span>
<span class="w">  </span><span class="nt">auto</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>from-bcl</strong>: Whether to convert BCL files to FASTQ files. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, the pipeline will use the <code class="docutils literal notranslate"><span class="pre">illumina-dir</span></code> path to find BCL files and convert them to FASTQ format. Only applicable for Illumina data.</p></li>
<li><p><strong>fastq.auto</strong>: If <code class="docutils literal notranslate"><span class="pre">True</span></code>, the pipeline expects FASTQ files to be in <code class="docutils literal notranslate"><span class="pre">resources/reads/</span></code> directory with the naming pattern <code class="docutils literal notranslate"><span class="pre">{sample_id}_1.fastq.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">{sample_id}_2.fastq.gz</span></code> for Illumina, or <code class="docutils literal notranslate"><span class="pre">{sample_id}.fastq.gz</span></code> for Nanopore. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, the pipeline expects the metadata file to have columns <code class="docutils literal notranslate"><span class="pre">fq1</span></code> and <code class="docutils literal notranslate"><span class="pre">fq2</span></code> (Illumina) or <code class="docutils literal notranslate"><span class="pre">fq1</span></code> (Nanopore) specifying the paths to the FASTQ files.</p></li>
</ul>
</section>
<section id="quality-control-configuration">
<h2>Quality Control Configuration<a class="headerlink" href="#quality-control-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">quality-control</span><span class="p">:</span>
<span class="w">  </span><span class="nt">sample-total-reads-threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">250</span>
<span class="w">  </span><span class="nt">amplicon-total-reads-threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>

<span class="w">  </span><span class="nt">coverage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">stats</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">multiqc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>quality-control.sample-total-reads-threshold</strong>: Minimum number of reads required for a sample to pass QC. Samples with fewer reads are removed at the quality control stage.</p></li>
<li><p><strong>quality-control.amplicon-total-reads-threshold</strong>: Minimum number of reads required for an amplicon to be considered for analysis.</p></li>
<li><p><strong>quality-control.coverage</strong>: Whether to calculate and report coverage statistics for each sample.</p></li>
<li><p><strong>quality-control.stats</strong>: Whether to generate alignment and variant calling statistics.</p></li>
<li><p><strong>quality-control.multiqc</strong>: Whether to generate a MultiQC report aggregating various QC metrics.</p></li>
</ul>
</section>
<section id="analysis-configuration">
<h2>Analysis Configuration<a class="headerlink" href="#analysis-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">analysis</span><span class="p">:</span><span class="w">  </span>
<span class="w">  </span><span class="nt">sample-map</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># needs lat and longs in metadata/sample_sheet</span>
<span class="w">  </span><span class="nt">population-structure</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">genetic-diversity</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">allele-frequencies</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>analysis.sample-map</strong>: Whether to generate a geographic map of sample collection locations. Requires latitude and longitude columns in the metadata file.</p></li>
<li><p><strong>analysis.population-structure</strong>: Whether to perform population structure analysis using principal component analysis (PCA).</p></li>
<li><p><strong>analysis.genetic-diversity</strong>: Whether to calculate genetic diversity metrics (such as nucleotide diversity).</p></li>
<li><p><strong>analysis.allele-frequencies</strong>: Whether to calculate and visualize allele frequencies across samples and groups.</p></li>
</ul>
</section>
<section id="jupyter-book-configuration">
<h2>Jupyter Book Configuration<a class="headerlink" href="#jupyter-book-configuration" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">build-jupyter-book</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>build-jupyter-book</strong>: Whether to compile all analysis notebooks into a Jupyter Book for convenient browsing of results. The book will be available at <code class="docutils literal notranslate"><span class="pre">results/ampseeker-results/_build/html/index.html</span></code>.</p></li>
</ul>
</section>
</section>
<section id="special-analyses-for-ag-vampir-panel">
<h1>Special Analyses for Ag-vampIR Panel<a class="headerlink" href="#special-analyses-for-ag-vampir-panel" title="Link to this heading">#</a></h1>
<p>When using the Ag-vampIR panel (<code class="docutils literal notranslate"><span class="pre">panel:</span> <span class="pre">ag-vampir</span></code>), the workflow automatically enables additional analyses:</p>
<ol class="arabic simple">
<li><p><strong>Species Identification</strong>: Identifies Anopheles species using amplicon sequencing data (results in <code class="docutils literal notranslate"><span class="pre">results/notebooks/ag-vampir/species-id.ipynb</span></code>).</p></li>
<li><p><strong>Kdr Analysis</strong>: Analyzes knockdown resistance mutations in the voltage-gated sodium channel (results in <code class="docutils literal notranslate"><span class="pre">results/notebooks/ag-vampir/kdr-analysis.ipynb</span></code>).</p></li>
</ol>
</section>
<section id="example-configuration">
<h1>Example Configuration<a class="headerlink" href="#example-configuration" title="Link to this heading">#</a></h1>
<p>Here’s a complete example configuration for reference:</p>
<section id="illumina-configuration-example">
<h2>Illumina Configuration Example<a class="headerlink" href="#illumina-configuration-example" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dataset and panel information</span>
<span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">lab-strains</span>
<span class="nt">panel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ag-vampir</span>
<span class="nt">platform</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">illumina</span>
<span class="nt">cohort-columns</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">location</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">taxon</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">country</span>
<span class="nt">targets</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/ag-vampir.bed</span>
<span class="nt">metadata</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/metadata.tsv</span>

<span class="c1"># Illumina directory (if using BCL files)</span>
<span class="nt">illumina-dir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/250110_M05658_0028_000000000-LTBV4</span>

<span class="c1"># Reference genome information</span>
<span class="nt">reference-name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AgamP4</span>
<span class="nt">reference-fasta</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/reference/Anopheles-gambiae-PEST_CHROMOSOMES_AgamP4.fa</span>
<span class="nt">reference-gff3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/reference/Anopheles-gambiae-PEST_BASEFEATURES_AgamP4.12.gff3</span>
<span class="nt">reference-snpeffdb</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Anopheles_gambiae</span>
<span class="nt">custom-snpeffdb</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>

<span class="c1"># Input file type options</span>
<span class="nt">from-bcl</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="nt">fastq</span><span class="p">:</span>
<span class="w">  </span><span class="nt">auto</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="c1"># Quality control options</span>
<span class="nt">quality-control</span><span class="p">:</span>
<span class="w">  </span><span class="nt">sample-total-reads-threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">250</span>
<span class="w">  </span><span class="nt">amplicon-total-reads-threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">  </span><span class="nt">fastp</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">coverage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">stats</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">multiqc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="c1"># Analysis options</span>
<span class="nt">analysis</span><span class="p">:</span><span class="w">  </span>
<span class="w">  </span><span class="nt">sample-map</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">  </span><span class="nt">population-structure</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">genetic-diversity</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">allele-frequencies</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="c1"># Build Jupyter book</span>
<span class="nt">build-jupyter-book</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
</section>
<section id="nanopore-configuration-example">
<h2>Nanopore Configuration Example<a class="headerlink" href="#nanopore-configuration-example" title="Link to this heading">#</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dataset and panel information</span>
<span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">nanopore-study</span>
<span class="nt">panel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">custom-panel</span>
<span class="nt">platform</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">nanopore</span>
<span class="nt">cohort-columns</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">location</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">taxon</span>
<span class="nt">targets</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/custom-panel.bed</span>
<span class="nt">metadata</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/metadata.tsv</span>

<span class="c1"># Reference genome information</span>
<span class="nt">reference-name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AgamP4</span>
<span class="nt">reference-fasta</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/reference/Anopheles-gambiae-PEST_CHROMOSOMES_AgamP4.fa</span>
<span class="nt">reference-gff3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/reference/Anopheles-gambiae-PEST_BASEFEATURES_AgamP4.12.gff3</span>
<span class="nt">reference-snpeffdb</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Anopheles_gambiae</span>
<span class="nt">custom-snpeffdb</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>

<span class="c1"># Input file type options (no BCL conversion for Nanopore)</span>
<span class="nt">from-bcl</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="nt">fastq</span><span class="p">:</span>
<span class="w">  </span><span class="nt">auto</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="c1"># Quality control options</span>
<span class="nt">quality-control</span><span class="p">:</span>
<span class="w">  </span><span class="nt">sample-total-reads-threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">250</span>
<span class="w">  </span><span class="nt">amplicon-total-reads-threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">  </span><span class="nt">fastp</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">coverage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">stats</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">multiqc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="c1"># Analysis options</span>
<span class="nt">analysis</span><span class="p">:</span><span class="w">  </span>
<span class="w">  </span><span class="nt">sample-map</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">  </span><span class="nt">population-structure</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">genetic-diversity</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">  </span><span class="nt">allele-frequencies</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="c1"># Build Jupyter book</span>
<span class="nt">build-jupyter-book</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="input_data.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Input files</p>
      </div>
    </a>
    <a class="right-next"
       href="running_the_workflow.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Running the workflow</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Configuring the workflow</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#configuration-parameters">Configuration Parameters</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cohort-analysis-configuration">Cohort Analysis Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-files-configuration">Input Files Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#illumina-directory-configuration">Illumina Directory Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-genome-configuration">Reference Genome Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-file-type-configuration">Input File Type Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-control-configuration">Quality Control Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis-configuration">Analysis Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#jupyter-book-configuration">Jupyter Book Configuration</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#special-analyses-for-ag-vampir-panel">Special Analyses for Ag-vampIR Panel</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example-configuration">Example Configuration</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#illumina-configuration-example">Illumina Configuration Example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#nanopore-configuration-example">Nanopore Configuration Example</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Sanjay Curtis Nagi
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>