
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Input files &#8212; AmpSeeker-documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css?v=6644e6bb" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "sanjaynagi/ampseeker");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/input_data';</script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Configuring the workflow" href="configuring_the_workflow.html" />
    <link rel="prev" title="Installation" href="installation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../landing-page.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="AmpSeeker-documentation - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="AmpSeeker-documentation - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../landing-page.html">
                    AmpSeeker
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Input files</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuring_the_workflow.html">Configuring the workflow</a></li>



<li class="toctree-l1"><a class="reference internal" href="running_the_workflow.html">Running the workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakemake.html">Snakemake</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting and contributing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to the workflow</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/sanjaynagi/ampseeker" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/sanjaynagi/ampseeker/issues/new?title=Issue%20on%20page%20%2Fnotebooks/input_data.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/input_data.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Input files</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sequencing-data-input">Sequencing Data Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#illumina-data">Illumina Data</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#samplesheet-csv-structure">SampleSheet.csv Structure</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#fastq-file-specifications-for-illumina">FASTQ File Specifications for Illumina</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#nanopore-data">Nanopore Data</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#fastq-file-specifications-for-nanopore">FASTQ File Specifications for Nanopore</a></li>
</ul>
</li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="input-files">
<h1>Input files<a class="headerlink" href="#input-files" title="Link to this heading">#</a></h1>
<p>AmpSeeker requires only:</p>
<ul class="simple">
<li><p>Sample metadata file <code class="docutils literal notranslate"><span class="pre">.tsv</span></code> or SampleSheet.csv in Illumina run folder containing metadata</p></li>
<li><p>Bed file of amplicon target SNP coordinates <code class="docutils literal notranslate"><span class="pre">.bed</span></code></p></li>
<li><p>Paired-end Illumina Amplicon-Seq data <code class="docutils literal notranslate"><span class="pre">.fastq.gz</span></code> OR single-end Nanopore amplicon data <code class="docutils literal notranslate"><span class="pre">.fastq.gz</span></code></p></li>
<li><p>Reference genome files</p></li>
</ul>
<p><strong>Sample metadata file</strong></p>
<p>If working directly from the Illumina BCL folder, a SampleSheet (<code class="docutils literal notranslate"><span class="pre">SampleSheet.csv</span></code>) must be placed in the Illumina run folder. An example sample sheet is located at <code class="docutils literal notranslate"><span class="pre">resources/exampleSampleSheet.csv</span></code>.</p>
<p>If working directly from fastq files, the user must provide either a SampleSheet or a sample metadata file in .tsv format which should be placed in the config/ folder and pointed to in the users <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. The default name is <code class="docutils literal notranslate"><span class="pre">metadata.tsv</span></code>. The file only requires one column, sample_id, and a column with which to group the data for downstream analyses (for example, by location). Any extra metadata columns can be included (e.g location, taxon, strain, country, year etc) and used to split data and colour the results figures, with configurable options in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file (cohort-columns).</p>
<p><strong>metadata.tsv:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>sample_id</p></th>
<th class="head"><p>location</p></th>
<th class="head"><p>taxon</p></th>
<th class="head"><p>latitude</p></th>
<th class="head"><p>longitude</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ContTia1</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
<td><p>7.1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>ContTia2</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
<td><p>7.1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>ContTia4</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
<td><p>7.1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>MalaTia1</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
<td><p>7.1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>MalaTia2</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
<td><p>7.1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>MalaTia4</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
<td><p>7.1</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
</div>
<p>Latitude and Longitudes are used to plot sample collection locations. If sample maps are not required, these columns can be omitted.</p>
<hr class="docutils" />
<p><strong>Bed file of amplicon targets</strong></p>
<p>Users should also provide a file in bed format, with at least 5 columns (chromosome, start, end, amplicon_id, target_label), and optionally reference and alternative alleles. The file should have <strong>no header</strong>. The bed file contains the coordinates of the amplicon targets and should be placed in the <code class="docutils literal notranslate"><span class="pre">config/</span></code> folder and pointed to in the users <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. The default name is <code class="docutils literal notranslate"><span class="pre">amplicon_targets.bed</span></code>.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>2L</p></th>
<th class="head"><p>209535</p></th>
<th class="head"><p>209536</p></th>
<th class="head"><p>Agam_1</p></th>
<th class="head"><p>AIM1</p></th>
<th class="head"><p>A</p></th>
<th class="head"><p>G</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2L</p></td>
<td><p>927246</p></td>
<td><p>927247</p></td>
<td><p>Agam_2</p></td>
<td><p>AIM2</p></td>
<td><p>C</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-odd"><td><p>2L</p></td>
<td><p>1274352</p></td>
<td><p>1274353</p></td>
<td><p>Agam_3</p></td>
<td><p>AIM3</p></td>
<td><p>G</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-even"><td><p>2L</p></td>
<td><p>1418209</p></td>
<td><p>1418210</p></td>
<td><p>Agam_4</p></td>
<td><p>AIM4</p></td>
<td><p>T</p></td>
<td><p>C</p></td>
</tr>
<tr class="row-odd"><td><p>2L</p></td>
<td><p>1571928</p></td>
<td><p>1571929</p></td>
<td><p>Agam_5</p></td>
<td><p>AIM5</p></td>
<td><p>T</p></td>
<td><p>C</p></td>
</tr>
</tbody>
</table>
</div>
<p>The columns are:</p>
<ol class="arabic simple">
<li><p><strong>Chromosome</strong>: Genomic chromosome or contig name (e.g., “2L”)</p></li>
<li><p><strong>Start position</strong>: 0-based start coordinate of the target</p></li>
<li><p><strong>End position</strong>: End coordinate of the target (typically start+1 for SNPs)</p></li>
<li><p><strong>Amplicon ID</strong>: Identifier for the amplicon (e.g., “Agam_1”)</p></li>
<li><p><strong>Target label</strong>: Descriptive name for the target (e.g., “AIM1”)</p></li>
<li><p><strong>Reference allele</strong>: (Optional) The reference allele at this position</p></li>
<li><p><strong>Alternate allele(s)</strong>: (Optional) The alternate allele(s) at this position</p></li>
</ol>
<p>The <code class="docutils literal notranslate"><span class="pre">target_label</span></code> column values should be unique for each row of the bed file. The <code class="docutils literal notranslate"><span class="pre">amplicon_id</span></code> column values can be repeated for multiple rows, to indicate that those SNPs are located on the same amplicon.</p>
<hr class="docutils" />
<section id="sequencing-data-input">
<h2>Sequencing Data Input<a class="headerlink" href="#sequencing-data-input" title="Link to this heading">#</a></h2>
<section id="illumina-data">
<h3>Illumina Data<a class="headerlink" href="#illumina-data" title="Link to this heading">#</a></h3>
<p><strong>Illumina run folders or Paired-end Amplicon-Sequencing fastq reads</strong></p>
<p>Users can either provide the path to an Illumina MiSeq run folder, or provide paired-end fastq reads. If converting BCL files to fastq within the workflow, a <code class="docutils literal notranslate"><span class="pre">SampleSheet.csv</span></code> must be placed in the Illumina run folder. The SampleSheet must have the CreateFastqForIndexReads parameter set to 1, please see the <code class="docutils literal notranslate"><span class="pre">exampleSampleSheet.csv</span></code> in the <code class="docutils literal notranslate"><span class="pre">resources/</span></code> directory.</p>
<section id="samplesheet-csv-structure">
<h4>SampleSheet.csv Structure<a class="headerlink" href="#samplesheet-csv-structure" title="Link to this heading">#</a></h4>
<p>The SampleSheet.csv is a structured file with several sections:</p>
<ul class="simple">
<li><p><strong>[Header]</strong>: Contains experiment metadata</p></li>
<li><p><strong>[Reads]</strong>: Defines read lengths (typically 151 for paired-end reads)</p></li>
<li><p><strong>[Settings]</strong>: Contains critical parameters including <code class="docutils literal notranslate"><span class="pre">CreateFastqForIndexReads=1</span></code></p></li>
<li><p><strong>[Data]</strong>: Sample information including IDs, indexes, and metadata</p></li>
</ul>
<p>The <code class="docutils literal notranslate"><span class="pre">CreateFastqForIndexReads=1</span></code> setting in the [Settings] section is essential as it instructs the BCL conversion to generate FASTQ files for index reads, which are used for demultiplexing samples.</p>
<p>Example [Data] section format:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>sample_ID</p></th>
<th class="head"><p>sample_name</p></th>
<th class="head"><p>index</p></th>
<th class="head"><p>index2</p></th>
<th class="head"><p>well</p></th>
<th class="head"><p>plate_name</p></th>
<th class="head"><p>taxon</p></th>
<th class="head"><p>location</p></th>
<th class="head"><p>country</p></th>
<th class="head"><p>latitude</p></th>
<th class="head"><p>longitude</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>GH_01</p></td>
<td><p>GH_01</p></td>
<td><p>ATCACGTT</p></td>
<td><p>CCTATCCT</p></td>
<td><p>A1</p></td>
<td><p>3</p></td>
<td><p></p></td>
<td><p>Obuasi</p></td>
<td><p>Ghana</p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p>GH_02</p></td>
<td><p>GH_02</p></td>
<td><p>CGATGTTT</p></td>
<td><p>CCTATCCT</p></td>
<td><p>A2</p></td>
<td><p>3</p></td>
<td><p></p></td>
<td><p>Obuasi</p></td>
<td><p>Ghana</p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
<p>Required columns for the [Data] section:</p>
<ul class="simple">
<li><p><strong>sample_ID</strong>: Unique sample identifier, used to name output files</p></li>
<li><p><strong>sample_name</strong>: Name displayed in reports (often the same as sample_ID)</p></li>
<li><p><strong>index</strong>: Forward index sequence for demultiplexing</p></li>
<li><p><strong>index2</strong>: Reverse index sequence for demultiplexing (for dual indexing)</p></li>
</ul>
<p>Optional metadata columns can be added (e.g., well, plate_name, taxon, location, country, latitude, longitude) and will be incorporated into analysis results.</p>
</section>
<section id="fastq-file-specifications-for-illumina">
<h4>FASTQ File Specifications for Illumina<a class="headerlink" href="#fastq-file-specifications-for-illumina" title="Link to this heading">#</a></h4>
<p>If providing fastq files, two gzipped fastq files for each sample are required, one for each pair of paired-end reads. Reads can be already trimmed or AmpSeeker can trim them, using the fastp module (configurable in the config.yaml).</p>
<p>Two options are available for specifying the location of the fastq files. Either the metadata contains two columns <code class="docutils literal notranslate"><span class="pre">fq1</span></code> and <code class="docutils literal notranslate"><span class="pre">fq2</span></code> with the paths to the fastq files, or the fastq files are placed in the following directory (<code class="docutils literal notranslate"><span class="pre">resources/reads/</span></code>) with the following naming pattern:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>ampseeker_dir/resources/reads/

Reads should be named as `{sample_id}_1.fastq.gz`, `{sample_id}_2.fastq.gz`.
</pre></div>
</div>
<p>If providing fastq paths in the metadata file, they can be named anything.</p>
</section>
</section>
<section id="nanopore-data">
<h3>Nanopore Data<a class="headerlink" href="#nanopore-data" title="Link to this heading">#</a></h3>
<section id="fastq-file-specifications-for-nanopore">
<h4>FASTQ File Specifications for Nanopore<a class="headerlink" href="#fastq-file-specifications-for-nanopore" title="Link to this heading">#</a></h4>
<p>For Nanopore data, provide single gzipped fastq files for each sample. The metadata file should contain a single column <code class="docutils literal notranslate"><span class="pre">fq1</span></code> with paths to the fastq files, or place files in <code class="docutils literal notranslate"><span class="pre">resources/reads/</span></code> with the naming pattern <code class="docutils literal notranslate"><span class="pre">{sample_id}.fastq.gz</span></code>.</p>
<p><strong>Nanopore metadata.tsv format:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>sample_id</p></th>
<th class="head"><p>fq1</p></th>
<th class="head"><p>location</p></th>
<th class="head"><p>taxon</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ContTia1</p></td>
<td><p>reads/ContTia1.fq.gz</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
</tr>
<tr class="row-odd"><td><p>ContTia2</p></td>
<td><p>reads/ContTia2.fq.gz</p></td>
<td><p>Tiassale</p></td>
<td><p>coluzzii</p></td>
</tr>
</tbody>
</table>
</div>
<p>Or with automatic file detection:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ampseeker_dir</span><span class="o">/</span><span class="n">resources</span><span class="o">/</span><span class="n">reads</span><span class="o">/</span>
<span class="n">ContTia1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">ContTia2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<hr class="docutils" />
<p><strong>Reference genome files</strong></p>
<p>For both platforms, genome alignment requires a fasta file containing the genome sequence. All input <code class="docutils literal notranslate"><span class="pre">.fa</span></code> files can be gzipped <code class="docutils literal notranslate"><span class="pre">.fa.gz</span></code>.</p>
<p>Reference genomes can be downloaded from vectorbase with the <code class="docutils literal notranslate"><span class="pre">resources/reference/download-vectorbase-reference.sh</span></code> script. Run it from the root AmpSeeker directory.</p>
<p>The user provides the path to the reference files in the configuration file (<code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>).</p>
<ol class="arabic simple">
<li><p><strong>Genome chromosomes reference file (.fa/.fa.gz)</strong>. Contains the DNA sequence for the genome in fasta format.</p></li>
<li><p><strong>Genome feature file (.gff3 format)</strong>.</p></li>
</ol>
<p>Ensure that contigs in the reference genome match the contigs in the bed file.</p>
<p>*Note - Genome reference files from VectorBase now have prefixes before each contig name, such as ‘AgamP4_2L’. Either the bed file can be updated to match these, or the names in the reference files can be modified. The downloader script can strip these prefixes automatically.</p>
</section>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="installation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Installation</p>
      </div>
    </a>
    <a class="right-next"
       href="configuring_the_workflow.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Configuring the workflow</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sequencing-data-input">Sequencing Data Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#illumina-data">Illumina Data</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#samplesheet-csv-structure">SampleSheet.csv Structure</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#fastq-file-specifications-for-illumina">FASTQ File Specifications for Illumina</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#nanopore-data">Nanopore Data</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#fastq-file-specifications-for-nanopore">FASTQ File Specifications for Nanopore</a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Sanjay Curtis Nagi
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>