{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c01954f",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import plotly.express as px\n",
    "\n",
    "sys.path.append(os.path.join(os.getcwd(), 'workflow/lib'))\n",
    "import ampseeker as amp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6302ed5",
   "metadata": {
    "tags": [
     "parameters",
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "dataset = 'nomads16'\n",
    "vcf_path = f\"../../results/vcfs/amplicons/{dataset}.annot.vcf\"\n",
    "metadata_path = \"../../results/config/metadata.qcpass.tsv\"\n",
    "cohort_cols = 'library_name'\n",
    "wkdir = \"../../\"\n",
    "platform = 'nanopore'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72254713",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append(os.path.join(wkdir, 'workflow/lib'))\n",
    "import ampseeker as amp\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "696938a0",
   "metadata": {},
   "source": [
    "## Population structure\n",
    "\n",
    "In this notebook, we run a principal components analysis and build a neighbour joining tree on the amplicon sequencing variant data. For the PCA, we will plot PC1 v PC2 and PC3 v PC4, and the variance explained by the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7ae6392",
   "metadata": {
    "tags": [
     "remove-input",
     "remove-output"
    ]
   },
   "outputs": [],
   "source": [
    "cohort_cols = cohort_cols.split(\",\")\n",
    "cohort_col = cohort_cols[0]\n",
    "\n",
    "metadata = pd.read_csv(metadata_path, sep=\"\\t\")\n",
    "\n",
    "import json\n",
    "with open(f\"{wkdir}/results/config/metadata_colours.json\", 'r') as f:\n",
    "    color_mapping = json.load(f)\n",
    "\n",
    "geno, pos, contig, metadata, ref, alt, ann = amp.load_variants(vcf_path, metadata, platform=platform, filter_indel=True)\n",
    "df_pca, model = amp.pca(geno=geno, metadata=metadata, n_components=4, missing_threshold=0.2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b018edf",
   "metadata": {},
   "source": [
    "#### Variance explained\n",
    "\n",
    "The variance explained shows the proportion of total variance in the dataset that is captured by each principal component. Higher values indicate more informative components. As a general rule of thumb, when the variance explained for each PC begins to flatten out, that is when the PCs are no longer informative."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "572a0d44",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "fig = px.bar(model.explained_variance_ratio_ , labels={\n",
    "                     \"value\": \"Variance Explained\",\n",
    "                     \"index\": \"Principal Component\",\n",
    "                 }, template='simple_white', height=250, width=600)\n",
    "fig.update_layout(showlegend=False)\n",
    "\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f9b98cf",
   "metadata": {},
   "source": [
    "### PCA\n",
    "\n",
    "Principal Component Analysis (PCA) is a dimensionality reduction technique that transforms high-dimensional genetic data into a smaller set of uncorrelated variables (Reich et al., 2008). It helps visualize population structure and genetic relationships between samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da0e807c",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "for coh in cohort_cols:\n",
    "    fig1 = amp.plot_pca(df_pca, x='PC1',y='PC2',z='PC3', colour_column=coh, cohort_columns=cohort_cols, dataset=dataset, color_mapping=color_mapping)\n",
    "    fig1.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82bbe3f6",
   "metadata": {},
   "source": [
    "## NJT\n",
    "\n",
    "Neighbor-Joining Tree (NJT) is a clustering method that reconstructs evolutionary relationships between samples based on genetic distances (Saitou & Nei, 1987). It creates a tree where genetically similar samples cluster together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f0977f1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "njt_figures = amp.run_njt_analysis(\n",
    "    geno=geno,\n",
    "    metadata=metadata,\n",
    "    cohort_cols=cohort_cols,\n",
    "    cohort_col=cohort_col,\n",
    "    color_mapping=color_mapping,\n",
    "    wkdir=wkdir,\n",
    ")\n",
    "\n",
    "for fig in njt_figures:\n",
    "    fig.show()\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}