{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebdcabe6",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "sys.path.append(os.path.join(os.getcwd(), 'workflow/lib'))\n",
    "import ampseeker as amp\n",
    "\n",
    "read_ann_field = amp.read_ann_field\n",
    "vcf_to_excel = amp.vcf_to_excel\n",
    "vcf_to_df = amp.vcf_to_df\n",
    "split_rows_with_multiple_alleles = amp.split_rows_with_multiple_alleles\n",
    "convert_genotype_to_alt_allele_count = amp.convert_genotype_to_alt_allele_count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ab3057",
   "metadata": {
    "tags": [
     "parameters",
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "dataset = \"nomads16\"\n",
    "wkdir = \"../..\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "612c170f",
   "metadata": {},
   "source": [
    "# SNP data\n",
    "\n",
    "In this notebook, we display the variant calling and annotation results as a pandas dataframe and save it as an excel spreadsheet for the user to explore or analyse further. If the DataFrame is too large to display, please use the .xlsx file. \n",
    "\n",
    "#### Target SNP data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "983bb2c8",
   "metadata": {
    "scrolled": false,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "vcf_df = vcf_to_excel(\n",
    "    vcf_path=f\"{wkdir}/results/vcfs/targets/{dataset}.annot.vcf\",\n",
    "    excel_path=f\"{wkdir}/results/vcfs/targets/{dataset}-snps.xlsx\",\n",
    "    convert_genotypes=True,\n",
    "    split_multiallelic=True,\n",
    ")\n",
    "pd.set_option(\"display.max_rows\", 1000, \"display.max_columns\", 1000)\n",
    "vcf_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ff0828c",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80f41f95",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "from IPython.display import display, Markdown\n",
    "display(Markdown(f'<a href=\"{wkdir}/results/vcfs/targets/{dataset}-snps.xlsx\">Target SNP data (.xlsx)</a>'))\n",
    "display(Markdown(f'<a href=\"{wkdir}/results/vcfs/targets/{dataset}.annot.vcf\">Target SNP data (.vcf)</a>'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "490f4ab3",
   "metadata": {},
   "source": [
    "#### Whole-amplicon SNP data\n",
    "\n",
    "For times sake, we do not write out whole-amplicon .xlsx by default - it is slow. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1735dc71",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "vcf_df = vcf_to_excel(\n",
    "    vcf_path=f\"{wkdir}/results/vcfs/amplicons/{dataset}.annot.vcf\",\n",
    "    excel_path=f\"{wkdir}/results/vcfs/amplicons/{dataset}-snps.xlsx\",\n",
    "    convert_genotypes=True,\n",
    "    split_multiallelic=True\n",
    ")\n",
    "vcf_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c0c4f01",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57f18b21",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "display(Markdown(f'<a href=\"{wkdir}/results/vcfs/amplicons/{dataset}-snps.xlsx\">Whole amplicon SNP data (.xlsx)</a>'))\n",
    "display(Markdown(f'<a href=\"{wkdir}/results/vcfs/amplicons/{dataset}.annot.vcf\">Whole amplicon SNP data (.vcf)</a>'))\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}