{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1548fbc",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907d8f16",
   "metadata": {
    "tags": [
     "remove-input",
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "dataset = 'test-agvampir'\n",
    "metadata_path = '../../../tests/snakemake/results/config/metadata.qcpass.tsv'\n",
    "kdr_marker_snps_path = '../../../resources/ag-vampir/Kdr_marker_SNPs.csv'\n",
    "vcf_path = f\"../../../tests/snakemake/results/vcfs/targets/{dataset}.annot.vcf\"\n",
    "cohort_cols = 'location,taxon'\n",
    "wkdir = \"../../../tests/snakemake/\"\n",
    "platform = 'illumina'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34463131",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import warnings\n",
    "\n",
    "sys.path.append(os.path.join(wkdir, 'workflow/lib'))\n",
    "import ampseeker as amp\n",
    "\n",
    "warnings.filterwarnings('ignore')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d83db822",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### *Kdr* origins and diplotype clustering\n",
    "\n",
    "The *Vgsc* gene encodes a voltage-gated sodium channel which is the binding target of DDT and pyrethroid insecticides. Variants in this gene, referred to as *Kdr* (knockdown resistance) are associated with resistance to DDT and/or pyrethroids.\n",
    "\n",
    "This notebook determines the origin of *Kdr* for each sample, and performs diplotype clustering to visualise the distribution of amino acids and haplotypes. \n",
    "\n",
    "Knockdown resistance mutations have arisen independently multiple times in different mosquito populations, and genetic backgrounds can help trace the origin and spread of resistance (Martinez-Torres et al., 1998; Ranson et al., 2000)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "257966e1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "cohort_cols = cohort_cols.split(\",\")\n",
    "cohort_col = cohort_cols[0]\n",
    "\n",
    "hap_def = pd.read_csv(kdr_marker_snps_path, sep = '\\t', index_col = 1)\n",
    "hap_def['variant_pos'] = hap_def.index.str.replace('.*:', '', regex = True).astype('int')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8beaef71",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Kdr background call functions extracted to workflow/lib/ag_vampir/kdr_analysis.py\n",
    "_F_kdr_origin_gen = amp._F_kdr_origin_gen\n",
    "_S_kdr_origin_gen = amp._S_kdr_origin_gen\n",
    "_402_kdr_origin_gen = amp._402_kdr_origin_gen\n",
    "_kdr_gen_cleanup = amp._kdr_gen_cleanup\n",
    "kdr_origin = amp.kdr_origin\n",
    "get_single_gen_call = amp.get_single_gen_call\n",
    "_get_single_gen_call_no_402 = amp._get_single_gen_call_no_402\n",
    "_get_single_gen_call_with_402 = amp._get_single_gen_call_with_402\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd1a611c",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Load metadata\n",
    "metadata = pd.read_csv(metadata_path, sep = '\\t')\n",
    "metadata = metadata.assign(taxon=metadata.taxon.fillna('UNKN'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5de2ea6",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Get the genotype calls\n",
    "### Filter the SNPs to just the ones useful for kdr origin analysis\n",
    "geno, pos, contig, metadata, ref, alt, ann = amp.load_vcf(vcf_path, metadata=metadata, platform=platform)\n",
    "samples = metadata.sample_id\n",
    "\n",
    "which_snps = (contig == '2L') & np.isin(pos, hap_def['variant_pos']) \n",
    "\n",
    "snp_calls = geno[which_snps, :, :]\n",
    "pos = pos[which_snps]\n",
    "alt = alt[which_snps, :]\n",
    "ref = ref[which_snps]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f235e86",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Convert genotype calls to nucleotides\n",
    "# combine ref and alt into a single matrix, and add a column of '?' at the end, so that \n",
    "# any genotype call of -1 (missing) draws the '?' character'\n",
    "snp_alleles = np.concatenate([np.reshape(ref, (len(ref), 1)), \n",
    "                              alt,\n",
    "                              np.full((len(ref), 1), '?')], \n",
    "                             axis = 1)\n",
    "\n",
    "# Convert numberic calls to nucleotides, and sort each pair of nucleotides alphabetically\n",
    "# (so, eg, the genotype 'TA' becomes 'AT')\n",
    "snp_genotypes_3d = snp_alleles[\n",
    "    np.array(np.arange(snp_alleles.shape[0])).reshape(snp_alleles.shape[0], 1, 1), \n",
    "    snp_calls\n",
    "]\n",
    "snp_genotypes = np.apply_along_axis(lambda x: ''.join(np.sort(x)), 2, snp_genotypes_3d)\n",
    "\n",
    "# Store results in data frame\n",
    "hap_def.index = hap_def['variant_pos']\n",
    "gen_df = pd.DataFrame(\n",
    "    np.transpose(snp_genotypes), \n",
    "    index = samples,\n",
    "    columns = hap_def.loc[pos, 'SNP name']\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8eb15c6",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Obtain kdr origin calls\n",
    "kdr_origins = pd.concat([kdr_origin(gen_df.iloc[i]) for i in range(gen_df.shape[0])])\n",
    "kdr_origins['kdr_origin'] = kdr_origins.apply(\n",
    "    get_single_gen_call, axis = 1\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f47c9a",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "#### Merge kdr origins with metadata and write to file. \n",
    "kdr_origins_df = pd.merge(kdr_origins, metadata.set_index(\"sample_id\"), left_index = True, right_index = True).rename_axis(\"sample_id\")\n",
    "kdr_origins_df.to_csv(f'{wkdir}/results/ag-vampir/kdr-origins/kdr_origins.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac2402e",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Create a table where each row is a haplotype instead of a genotype (although these are genotype-based calls, \n",
    "# so order within each sample will be random, so they can be used for, say, mapping, but not haplotype clustering). Write this table to file.\n",
    "kdr_genhap_origins_df = pd.DataFrame({'kdr_origin': ','.join(list(kdr_origins['kdr_origin'])).split(',')},\n",
    "                                     index = np.repeat(kdr_origins.index, 2)\n",
    ")\n",
    "kdr_genhap_origins_df = pd.merge(kdr_genhap_origins_df, metadata.set_index(\"sample_id\"), left_index=True, right_index=True)\n",
    "kdr_genhap_origins_df.to_csv(f'{wkdir}/results/ag-vampir/kdr-origins/kdr_genhap_origins.tsv', sep = '\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3b20e5",
   "metadata": {
    "scrolled": false,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "cols_keep = cohort_cols + ['kdr_origin']\n",
    "# Count the number of occurances of each haplotypes in each population\n",
    "# \"values\" could be any column that isn't specified elsewhere in the function. But it's \n",
    "# not allowed to be blank, so we had to pick one. \n",
    "pop_origin_counts = kdr_genhap_origins_df[cols_keep].pivot_table(columns='kdr_origin', \n",
    "                                                                    index=cohort_cols,\n",
    "                                                                    aggfunc=len\n",
    "                                                                   ).fillna(0).astype(int)\n",
    "pop_origin_counts.to_excel(f'{wkdir}/results/ag-vampir/kdr-origins/kdr_origin_counts.xlsx')\n",
    "# A function to round a number up to n_signif significant figures\n",
    "signif = amp.signif\n",
    "\n",
    "# Calculate row totals of non-\"?\" columns\n",
    "if '?' in pop_origin_counts.columns:\n",
    "    pop_origin_counts = pop_origin_counts.drop('?', axis = 1)\n",
    "\n",
    "row_totals = pop_origin_counts.sum(axis = 1)\n",
    "# Calculate origin frequencies. We exclude the \"?\" calls for this\n",
    "pop_origin_freqs = pop_origin_counts.div(row_totals, axis = 0)\n",
    "pop_origin_freqs.to_excel(f'{wkdir}/results/ag-vampir/kdr-origins/kdr_origin_freqs.xlsx')\n",
    "# Round to 2 significant figures\n",
    "pop_origin_freqs = signif(pop_origin_freqs, 2)\n",
    "print('Counts of origins:')\n",
    "display(pop_origin_counts)\n",
    "print('\\n\\nFrequencies of known origins:')\n",
    "display(pop_origin_freqs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a6e5a9a",
   "metadata": {},
   "source": [
    "#### Diplotype clustering\n",
    "\n",
    "Diplotype clustering groups samples based on genetic similarity at the Vgsc locus, revealing evolutionary relationships at resistance loci and helping identify mutations associated with selective sweeps (Nagi et al., 2024)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb0482d1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "\n",
    "_dipclust_concat_subplots = amp._dipclust_concat_subplots\n",
    "plot_dendrogram = amp.plot_dendrogram\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f53cf42",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "df_samples = pd.read_csv(metadata_path, sep=\"\\t\")\n",
    "df_kdr = pd.read_csv(f'{wkdir}/results/ag-vampir/kdr-origins/kdr_origins.tsv', sep=\"\\t\", index_col=0)\n",
    "df_kdr = df_kdr.reset_index().rename(columns={'level_0': 'sample_id'})\n",
    "\n",
    "vcf_path = f\"{wkdir}/results/vcfs/amplicons/{dataset}.annot.vcf\"\n",
    "geno, pos, contig, df_samples, ref, alt, ann = amp.load_vcf(vcf_path, metadata=df_samples, platform=platform)\n",
    "\n",
    "import json\n",
    "with open(f\"{wkdir}/results/config/metadata_colours.json\", 'r') as f:\n",
    "    color_mapping = json.load(f)\n",
    "\n",
    "bed_path = f\"{wkdir}/config/ag-vampir.bed\"\n",
    "df_bed = amp.load_bed(bed_path)\n",
    "\n",
    "# subset to VGSC SNPs\n",
    "vgsc_mask = df_bed.eval('mutation.str.contains(\"Vgsc\")').to_numpy()\n",
    "df_vgsc = df_bed[vgsc_mask]\n",
    "vgsc_start = df_vgsc['start'].min() - 200 \n",
    "vgsc_end = df_vgsc['end'].max() + 200\n",
    "vgsc_mask = np.logical_and(contig == '2L', np.logical_and(pos >= vgsc_start, pos <= vgsc_end))\n",
    "geno_vgsc = geno.compress(vgsc_mask, axis=0)\n",
    "pos_vgsc = pos[vgsc_mask]\n",
    "\n",
    "# remove invariant sites \n",
    "ac = geno_vgsc.count_alleles()\n",
    "is_seg = ac.is_segregating()\n",
    "geno_vgsc = geno_vgsc.compress(is_seg, axis=0)\n",
    "\n",
    "# remove highly missing sites \n",
    "missing_mask = geno_vgsc.is_missing().mean(axis=1) > 0.1\n",
    "missing_mask.sum()\n",
    "geno_vgsc = geno_vgsc.compress(~missing_mask, axis=0)\n",
    "\n",
    "# distances \n",
    "from scipy.spatial.distance import squareform\n",
    "\n",
    "ac = allel.GenotypeArray(geno_vgsc).to_allele_counts(max_allele=3)\n",
    "X = np.ascontiguousarray(np.swapaxes(ac.values, 0, 1))\n",
    "dists = amp.multiallelic_diplotype_pdist(X, metric=amp.multiallelic_diplotype_mean_cityblock)\n",
    "dist_matrix = squareform(dists)\n",
    "df_dists = pd.DataFrame(dist_matrix, index=df_samples.index, columns=df_samples.index)\n",
    "na_mask = df_dists.isna().any()\n",
    "df_dists = df_dists.loc[~na_mask, ~na_mask]\n",
    "\n",
    "df_samples = df_samples.reset_index().merge(df_kdr[['sample_id', 'kdr_origin']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e9d2cb",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import scipy\n",
    "\n",
    "distance_metric = 'cityblock'\n",
    "leaf_color = cohort_col\n",
    "\n",
    "fig_dendro, leaf_data = plot_dendrogram(\n",
    "    dist=scipy.spatial.distance.squareform(df_dists.values),\n",
    "    linkage_method=\"complete\",\n",
    "    count_sort=True,\n",
    "    distance_sort=False,\n",
    "    render_mode=\"svg\",\n",
    "    width=800,\n",
    "    height=500,\n",
    "    title=f\"{dataset} | Vgsc diplotype clustering\",\n",
    "    line_width=0.6,\n",
    "    line_color='black',\n",
    "    marker_size=5,\n",
    "    leaf_data=df_samples[~na_mask.to_numpy()],\n",
    "    leaf_hover_name=\"sample_id\",\n",
    "    leaf_hover_data=cohort_cols + ['kdr_origin'],\n",
    "    leaf_color=leaf_color,\n",
    "    leaf_symbol=None,\n",
    "    leaf_y=-0.01,\n",
    "    leaf_color_discrete_map=color_mapping[leaf_color],\n",
    "    leaf_category_orders=None,\n",
    "    template=\"simple_white\",\n",
    "    y_axis_title=f\"Distance ({distance_metric})\",\n",
    "    y_axis_buffer=0.1,\n",
    ")\n",
    "\n",
    "\n",
    "df_snps = pd.read_excel(f\"{wkdir}/results/vcfs/targets/{dataset}-snps.xlsx\")\n",
    "df_vgsc_snps = df_snps.query(\"CHROM == '2L' and POS >= @vgsc_start and POS <= @vgsc_end\")\n",
    "df_vgsc_snps = df_bed.rename(columns={'end':'POS', 'contig':'CHROM'})[['CHROM', 'POS', 'mutation']].merge(df_vgsc_snps, on=['CHROM', 'POS'])\n",
    "df_vgsc_snps = df_vgsc_snps.set_index('mutation').iloc[:, 6:]\n",
    "df_vgsc_snps = df_vgsc_snps.query(\"~mutation.str.contains('AIM')\").query(\"~mutation.str.contains('tag')\")\n",
    "df_vgsc_snps = df_vgsc_snps.loc[:, leaf_data.sample_id.to_list()]\n",
    "df_vgsc_snps.index = df_vgsc_snps.index.str.replace(\"Vgsc_\", \"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affe0b60",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "kdr_leaf_data = leaf_data.copy()\n",
    "\n",
    "figures = [fig_dendro]\n",
    "subplot_heights = [300]\n",
    "snp_row_height  = 20\n",
    "width = 800\n",
    "\n",
    "kdr_fig = px.scatter(\n",
    "    data_frame=kdr_leaf_data,\n",
    "    x=np.arange(kdr_leaf_data.shape[0]),\n",
    "    y=np.repeat(0, kdr_leaf_data.shape[0]),\n",
    "    color='kdr_origin',\n",
    "    hover_name='sample_id',\n",
    "    symbol_sequence=['square'],\n",
    "    # hover_data=leaf_hover_data,\n",
    "    template='simple_white',\n",
    "    # color_discrete_map=leaf_color_discrete_map,\n",
    "    # category_orders=leaf_category_orders,\n",
    ")\n",
    "\n",
    "for f in kdr_fig.data:\n",
    "    f.legendgroup = 'kdr_origin'\n",
    "\n",
    "figures.append(kdr_fig)\n",
    "subplot_heights.append(20)\n",
    "\n",
    "# het bar\n",
    "df_het = pd.DataFrame(\n",
    "    {\"sample_id\": samples, \"Sample Heterozygosity\": geno_vgsc.is_het().sum(axis=0) / geno_vgsc.is_called().sum(axis=0)}\n",
    ").set_index(\"sample_id\")\n",
    "\n",
    "# order according to dendrogram and transpose\n",
    "df_het = df_het.loc[leaf_data.sample_id, :].T\n",
    "het_trace = go.Heatmap(\n",
    "    z=df_het,\n",
    "    y=[\"Heterozygosity\"],\n",
    "    colorscale=\"Greys\",\n",
    "    showlegend=False,\n",
    "    showscale=False,\n",
    ")\n",
    "\n",
    "figures.append(het_trace)\n",
    "subplot_heights.append(30)\n",
    "\n",
    "snp_trace = go.Heatmap(\n",
    "    z=df_vgsc_snps[::-1].values,\n",
    "    y=df_vgsc_snps[::-1].index.to_list(),\n",
    "    colorscale=\"Greys\",\n",
    "    showlegend=False,\n",
    "    showscale=False,\n",
    ")\n",
    "\n",
    "figures.append(snp_trace)\n",
    "subplot_heights.append(snp_row_height * df_vgsc_snps.shape[0])\n",
    "\n",
    "height = sum(subplot_heights) + 50\n",
    "fig = _dipclust_concat_subplots(\n",
    "    figures=figures,\n",
    "    width=width,\n",
    "    height=height,\n",
    "    row_heights=subplot_heights,\n",
    "    title=f\"{dataset} | Vgsc diplotype clustering\",\n",
    "    xaxis_range=(0, df_dists.shape[0]),\n",
    ")\n",
    "\n",
    "fig[\"layout\"][\"yaxis\"][\"title\"] = f\"Distance ({distance_metric})\"\n",
    "\n",
    "aa_idx = len(figures)\n",
    "fig.add_hline(y=-0.5, line_width=1, line_color=\"grey\", row=aa_idx, col=1)\n",
    "for i, y in enumerate(df_vgsc_snps.index.to_list()):\n",
    "    fig.add_hline(y=i+0.5, line_width=1, line_color=\"grey\", row=aa_idx, col=1)\n",
    "\n",
    "fig['layout'][f'yaxis{aa_idx}']['title']=f'Vgsc mutations'\n",
    "fig.update_xaxes(showline = True, linecolor = 'grey', linewidth = 1, row = aa_idx, col = 1, mirror = True)\n",
    "fig.update_yaxes(showline = True, linecolor = 'grey', linewidth = 1, row = aa_idx, col = 1, mirror = True)\n",
    "fig.write_image(f\"{wkdir}/results/kdr_dipclust.png\", scale=2)\n",
    "fig.show()\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}