{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1548fbc",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907d8f16",
   "metadata": {
    "tags": [
     "remove-input",
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "dataset = 'ag-vampir-002'\n",
    "metadata_path = '../../../results/config/metadata.qcpass.tsv'\n",
    "kdr_marker_snps_path = '../../../resources/ag-vampir/Kdr_marker_SNPs.csv'\n",
    "vcf_path = f\"../../../results/vcfs/targets/{dataset}.annot.vcf\"\n",
    "cohort_cols = 'location,taxon'\n",
    "wkdir = \"../../..\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34463131",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.join(wkdir, '/workflow'))\n",
    "import ampseekertools as amp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d83db822",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### *Kdr* origins and diplotype clustering\n",
    "\n",
    "This notebook determines the origin of Kdr for each sample, and performs diplotype clustering.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "257966e1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "cohort_cols = cohort_cols.split(\",\")\n",
    "\n",
    "hap_def = pd.read_csv(kdr_marker_snps_path, sep = '\\t', index_col = 1)\n",
    "hap_def['variant_pos'] = hap_def.index.str.replace('.*:', '', regex = True).astype('int')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8beaef71",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "#### The functions for making the kdr background calls\n",
    "# Determine kdr F origin for a genotype\n",
    "def _F_kdr_origin_gen(genotypes, clean = True):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995F mutations\n",
    "    if pd.isnull(genotypes['kdr-995F']):\n",
    "        kdr_F_origins = 'F:unknown'\n",
    "    elif genotypes['kdr-995F'] == 'AA':\n",
    "        kdr_F_origins = 'F:wt_hom'\n",
    "    elif genotypes['kdr-995F'] == 'AT':\n",
    "        kdr_F_origins = 'F:het'\n",
    "    elif genotypes['kdr-995F'] == 'TT':\n",
    "        kdr_F_origins = 'F:hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr F genotype. {sample_name} {genotypes[\"kdr-995F\"]}')\n",
    "        kdr_F_origins = 'Fail. Unexpected kdr F genotype'\n",
    "    # If the individual has Fkdr, find out its origins\n",
    "    # For F homozygotes\n",
    "    if kdr_F_origins == 'F:hom':\n",
    "        if pd.isnull(genotypes['Def-F1']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1?'\n",
    "        elif genotypes['Def-F1'] == 'AA':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_hom'\n",
    "        elif genotypes['Def-F1'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2?'\n",
    "        elif genotypes['Def-F2'] == 'AA':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_hom'\n",
    "        elif genotypes['Def-F2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F3F4-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F3F4?'\n",
    "        elif genotypes['Def-F3F4-2'] == 'TT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_hom'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_hom'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het,F4_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_hom'\n",
    "        elif genotypes['Def-F3F4-2'] == 'AT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_het'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F3F4, but homozygote for F3.'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F5-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5?'\n",
    "        elif genotypes['Def-F5-2'] == 'GG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_hom'\n",
    "        elif genotypes['Def-F5-2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_het'\n",
    "    # for F heterozygotes\n",
    "    elif kdr_F_origins == 'F:het':\n",
    "        if pd.isnull(genotypes['Def-F1']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1?'\n",
    "        elif genotypes['Def-F1'] == 'AA':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F1.'\n",
    "        elif genotypes['Def-F1'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2?'\n",
    "        elif genotypes['Def-F2'] == 'AA':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F2.'\n",
    "        elif genotypes['Def-F2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F3F4-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F3F4?'\n",
    "        elif genotypes['Def-F3F4-2'] == 'TT':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F3F4.'\n",
    "        elif genotypes['Def-F3F4-2'] == 'AT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_het'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr and F3F4, but homozygote for F3.'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F5-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5?'\n",
    "        elif genotypes['Def-F5-2'] == 'GG':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F5.'\n",
    "        elif genotypes['Def-F5-2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_het'\n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_F_origins))\n",
    "    else:\n",
    "        return(kdr_F_origins)\n",
    "\n",
    "\n",
    "# Determine kdr S origin for a genotype\n",
    "def _S_kdr_origin_gen(genotypes, clean = True, alternate_S4S5 = False):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995S mutations\n",
    "    if pd.isnull(genotypes['kdr-995S']):\n",
    "        kdr_S_origins = 'S:unknown'\n",
    "    elif genotypes['kdr-995S'] == 'TT':\n",
    "        kdr_S_origins = 'S:wt_hom'\n",
    "    elif genotypes['kdr-995S'] == 'CT':\n",
    "        kdr_S_origins = 'S:het'\n",
    "    elif genotypes['kdr-995S'] == 'CC':\n",
    "        kdr_S_origins = 'S:hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr S genotype. {sample_name} {genotypes[\"kdr-995S\"]}')\n",
    "        kdr_S_origins = 'Fail. Unexpected kdr S genotype'\n",
    "    # If the individual has Skdr, find out its origins\n",
    "    # For S homozygotes\n",
    "    if kdr_S_origins == 'S:hom':\n",
    "        if pd.isnull(genotypes['Def-S1-3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1?'\n",
    "        elif genotypes['Def-S1-3'] == 'CC':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_hom'\n",
    "        elif genotypes['Def-S1-3'] == 'CT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S2S4']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S2S4?'\n",
    "        elif genotypes['Def-S2S4'] == 'TT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_hom'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_hom'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het,S4_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "        elif genotypes['Def-S2S4'] == 'CT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_het'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S2S4, but homozygote for S2.'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3?'\n",
    "        elif genotypes['Def-S3'] == 'GG':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_hom'\n",
    "        elif genotypes['Def-S3'] == 'GT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_het'\n",
    "        # \n",
    "        if alternate_S4S5:\n",
    "            if pd.isnull(genotypes['Def-S4S5-2']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5-2'] == 'TT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_hom'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_hom'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het,S4_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "            elif genotypes['Def-S4S5-2'] == 'GT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        else :\n",
    "            if pd.isnull(genotypes['Def-S4S5']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5'] == 'CC':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_hom'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_hom'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het,S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "            elif genotypes['Def-S4S5'] == 'CT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "    # for S heterozygotes\n",
    "    elif kdr_S_origins == 'S:het':\n",
    "        if pd.isnull(genotypes['Def-S1-3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1?'\n",
    "        elif genotypes['Def-S1-3'] == 'CC':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S1.'\n",
    "        elif genotypes['Def-S1-3'] == 'CT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S2S4']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S2S4?'\n",
    "        elif genotypes['Def-S2S4'] == 'TT':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S2S4.'\n",
    "        elif genotypes['Def-S2S4'] == 'CT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_het'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                ksr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S2S4, but homozygote for S2.'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3?'\n",
    "        elif genotypes['Def-S3'] == 'GG':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S3.'\n",
    "        elif genotypes['Def-S3'] == 'GT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_het'\n",
    "        # \n",
    "        if alternate_S4S5:\n",
    "            if pd.isnull(genotypes['Def-S4S5_2']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5-2'] == 'TT':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S4S5.'\n",
    "            elif genotypes['Def-S4S5-2'] == 'GT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        else :\n",
    "            if pd.isnull(genotypes['Def-S4S5']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5'] == 'CC':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S4S5.'\n",
    "            elif genotypes['Def-S4S5'] == 'CT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    return(f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S4S5, but homozygote for S5.')\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_S_origins))\n",
    "    else:\n",
    "        return(kdr_S_origins)\n",
    "\n",
    "def _402_kdr_origin_gen(genotypes, clean = True):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995F mutations\n",
    "    if pd.isnull(genotypes['kdr-402L']):\n",
    "        kdr_402_origins = '402:unknown'\n",
    "    elif genotypes['kdr-402L'] == 'GG':\n",
    "        kdr_402_origins = '402:wt_hom'\n",
    "    elif genotypes['kdr-402L'] == 'CG':\n",
    "        kdr_402_origins = '402:het,402LC_het'\n",
    "    elif genotypes['kdr-402L'] == 'GT':\n",
    "        kdr_402_origins = '402:het,402LT_het'\n",
    "    elif genotypes['kdr-402L'] == 'CT':\n",
    "        kdr_402_origins = '402:hom,402LC_het,402LT_het'\n",
    "    elif genotypes['kdr-402L'] == 'CC':\n",
    "        kdr_402_origins = '402:hom,402LC_hom'\n",
    "    elif genotypes['kdr-402L'] == 'TT':\n",
    "        kdr_402_origins = '402:hom,402LT_hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr 402 genotype. {sample_name} {genotypes[\"kdr-402L\"]}')\n",
    "        kdr_402_origins = 'Fail. Unexpected kdr 402 genotype'\n",
    "        \n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_402_origins))\n",
    "    else:\n",
    "        return(kdr_402_origins)\n",
    "\n",
    "# The initial output of the kdr_origin function can be a little messy, since it outputs all of the \n",
    "# information that it could or couldn't obtain. This function tidies it up a bit. \n",
    "def _kdr_gen_cleanup(kdr_origin_str):\n",
    "    if re.search('Fail', kdr_origin_str):\n",
    "        return('?,?')\n",
    "    if re.search('wt', kdr_origin_str):\n",
    "        return('wt,wt')\n",
    "    kdr_type = re.findall('.*(?=:)', kdr_origin_str)[0]\n",
    "    outcomes = kdr_origin_str.split(',')\n",
    "    origins = np.unique(outcomes[1:])\n",
    "    established_origins = [o for o in origins if not re.search(r'\\?', o)]\n",
    "    # Remove \"_het\" and \"_hom\" text\n",
    "    established_origins = [re.sub('_het', '', o) for o in established_origins]\n",
    "    # Now double up each _hom entry. this is cludgy, but coudn't find a more elegant way\n",
    "    for i in range(len(established_origins)):\n",
    "        o = established_origins[i]\n",
    "        if re.search('_hom', o):\n",
    "            o = re.sub('_hom', '', o)\n",
    "            established_origins[i] = o\n",
    "            established_origins.append(o)\n",
    "    if re.search('hom', outcomes[0]):\n",
    "        if len(established_origins) == 1:\n",
    "            return(f'{kdr_type},{established_origins[0]}')\n",
    "        elif len(established_origins) == 2:\n",
    "            return(','.join(established_origins))\n",
    "        else:\n",
    "            return(f'{kdr_type},{kdr_type}')\n",
    "    if re.search('het', outcomes[0]):\n",
    "        if len(established_origins) == 1:\n",
    "            return(f'wt,{established_origins[0]}')\n",
    "        else:\n",
    "            return(f'wt,{kdr_type}')\n",
    "    else:\n",
    "        return('?,?')\n",
    "\n",
    "\n",
    "# Single function to call both the F and S origins for a given haplotype. This function determines\n",
    "# from the look of the genotype table whether it represents genotypes or haplotypes, and calls \n",
    "# the appropriate function. \n",
    "def kdr_origin(genotypes, alternate_S4S5 = False, clean = True, include_402 = None):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    if include_402 == None:\n",
    "        if 'kdr-402L' in genotypes.index:\n",
    "            include_402 = True\n",
    "        else:\n",
    "            include_402 = False\n",
    "    if include_402 == False:\n",
    "        kdr_origins = pd.DataFrame({'kdr_F_origin': [_F_kdr_origin_gen(genotypes, clean)], \n",
    "                                    'kdr_S_origin': [_S_kdr_origin_gen(genotypes, clean, alternate_S4S5)]\n",
    "                                    }, index = [sample_name]\n",
    "        \n",
    "        )\n",
    "    else:\n",
    "        kdr_origins = pd.DataFrame({'kdr_F_origin': [_F_kdr_origin_gen(genotypes, clean)], \n",
    "                                    'kdr_S_origin': [_S_kdr_origin_gen(genotypes, clean, alternate_S4S5)],\n",
    "                                    'kdr_402_origin': [_402_kdr_origin_gen(genotypes, clean)]\n",
    "                                    }, index = [sample_name]\n",
    "        )\n",
    "    return(kdr_origins)\n",
    "\n",
    "# From a pair of kdr origin calls (obtained by running the kdr_origin function  for each of \n",
    "# F and S, followed by kdr_hap_cleanup), output a single call combining the F and S calls. \n",
    "def get_single_gen_call(x):  \n",
    "    if 'kdr_402_origin' in x.index:\n",
    "        return(_get_single_gen_call_with_402(x))\n",
    "    else:\n",
    "        return(_get_single_gen_call_no_402(x))\n",
    "    \n",
    "def _get_single_gen_call_no_402(x): \n",
    "    if 'sample_name' in x.index:\n",
    "        sample_name = x['sample_name']\n",
    "    else:\n",
    "        sample_name = x.name\n",
    "    joined_calls = np.array(x['kdr_F_origin'].split(',') + x['kdr_S_origin'].split(','))\n",
    "    # There should be at least two 'wt' calls\n",
    "    if np.sum(joined_calls == 'wt') < 2:\n",
    "        print(f'Too many different mutant haplotype backgrounds in sample {sample_name}')\n",
    "        return('?,?')\n",
    "    # Otherwise, drop two wildtype calls\n",
    "    else:\n",
    "        which_drop = np.where(joined_calls == 'wt')[0][:2]\n",
    "        return(','.join(np.delete(joined_calls, which_drop)))\n",
    "\n",
    "def _get_single_gen_call_with_402(x): \n",
    "    if 'sample_name' in x.index:\n",
    "        sample_name = x['sample_name']\n",
    "    else:\n",
    "        sample_name = x.name\n",
    "    joined_calls = np.array(x['kdr_F_origin'].split(',') + \n",
    "                            x['kdr_S_origin'].split(',') +\n",
    "                            x['kdr_402_origin'].split(',')\n",
    "    )\n",
    "    # There should be at least four 'wt' calls\n",
    "    if np.sum(np.isin(joined_calls,  ['wt', '?'])) < 4:\n",
    "        print(f'Too many different mutant haplotype backgrounds in sample {sample_name}')\n",
    "        return('?,?')\n",
    "    # Otherwise, drop four wildtype calls\n",
    "    else:\n",
    "        which_drop = np.concatenate([\n",
    "            np.where(joined_calls == 'wt')[0],\n",
    "            np.where(joined_calls == '?')[0]\n",
    "        ])[:4]\n",
    "        return(','.join(np.delete(joined_calls, which_drop)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd1a611c",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Load metadata\n",
    "metadata = pd.read_csv(metadata_path, sep = '\\t')\n",
    "metadata = metadata.assign(taxon=metadata.taxon.fillna('UNKN'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5de2ea6",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Get the genotype calls\n",
    "### Filter the SNPs to just the ones useful for kdr origin analysis\n",
    "geno, pos, contig, metadata, ref, alt = amp.load_vcf(vcf_path, metadata=metadata)\n",
    "samples = metadata.sample_id\n",
    "\n",
    "which_snps = (contig == '2L') & np.isin(pos, hap_def['variant_pos']) \n",
    "\n",
    "snp_calls = geno[which_snps, :, :]\n",
    "pos = pos[which_snps]\n",
    "alt = alt[which_snps, :]\n",
    "ref = ref[which_snps]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f235e86",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Convert genotype calls to nucleotides\n",
    "# combine ref and alt into a single matrix, and add a column of '?' at the end, so that \n",
    "# any genotype call of -1 (missing) draws the '?' character'\n",
    "snp_alleles = np.concatenate([np.reshape(ref, (len(ref), 1)), \n",
    "                              alt,\n",
    "                              np.full((len(ref), 1), '?')], \n",
    "                             axis = 1)\n",
    "\n",
    "# Convert numberic calls to nucleotides, and sort each pair of nucleotides alphabetically\n",
    "# (so, eg, the genotype 'TA' becomes 'AT')\n",
    "snp_genotypes_3d = snp_alleles[\n",
    "    np.array(np.arange(snp_alleles.shape[0])).reshape(snp_alleles.shape[0], 1, 1), \n",
    "    snp_calls\n",
    "]\n",
    "snp_genotypes = np.apply_along_axis(lambda x: ''.join(np.sort(x)), 2, snp_genotypes_3d)\n",
    "\n",
    "# Store results in data frame\n",
    "hap_def.index = hap_def['variant_pos']\n",
    "gen_df = pd.DataFrame(\n",
    "    np.transpose(snp_genotypes), \n",
    "    index = samples,\n",
    "    columns = hap_def.loc[pos, 'SNP name']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8eb15c6",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Obtain kdr origin calls\n",
    "kdr_origins = pd.concat([kdr_origin(gen_df.iloc[i]) for i in range(gen_df.shape[0])])\n",
    "kdr_origins['kdr_origin'] = kdr_origins.apply(\n",
    "    get_single_gen_call, axis = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f47c9a",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "#### Merge kdr origins with metadata and write to file. \n",
    "kdr_origins_df = pd.merge(kdr_origins, metadata.set_index(\"sample_id\"), left_index = True, right_index = True)\n",
    "kdr_origins_df.to_csv(f'{wkdir}/results/kdr-origins/kdr_origins.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac2402e",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "### Create a table where each row is a haplotype instead of a genotype (although these are genotype-based calls, \n",
    "# so order within each sample will be random, so they can be used for, say, mapping, but not haplotype clustering). Write this table to file.\n",
    "kdr_genhap_origins_df = pd.DataFrame({'kdr_origin': ','.join(list(kdr_origins['kdr_origin'])).split(',')},\n",
    "                                     index = np.repeat(kdr_origins.index, 2)\n",
    ")\n",
    "kdr_genhap_origins_df = pd.merge(kdr_genhap_origins_df, metadata.set_index(\"sample_id\"), left_index=True, right_index=True)\n",
    "kdr_genhap_origins_df.to_csv(f'{wkdir}/results/kdr-origins/kdr_genhap_origins.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3b20e5",
   "metadata": {
    "scrolled": false,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "cols_keep = cohort_cols + ['kdr_origin']\n",
    "# Count the number of occurances of each haplotypes in each population\n",
    "# \"values\" could be any column that isn't specified elsewhere in the function. But it's \n",
    "# not allowed to be blank, so we had to pick one. \n",
    "pop_origin_counts = kdr_genhap_origins_df[cols_keep].pivot_table(columns='kdr_origin', \n",
    "                                                                    index=cohort_cols,\n",
    "                                                                    aggfunc=len\n",
    "                                                                   ).fillna(0).astype(int)\n",
    "\n",
    "# A function to round a number up to n_signif significant figures\n",
    "def signif(x, n_figs):\n",
    "    power = 10 ** np.floor(np.log10(np.abs(x).clip(1e-200)))\n",
    "    rounded = np.round(x / power, n_figs - 1) * power\n",
    "    return rounded\n",
    "\n",
    "# Calculate row totals of non-\"?\" columns\n",
    "if '?' in pop_origin_counts.columns:\n",
    "    pop_origin_counts = pop_origin_counts.drop('?', axis = 1)\n",
    "\n",
    "row_totals = pop_origin_counts.sum(axis = 1)\n",
    "# Calculate origin frequencies. We exclude the \"?\" calls for this\n",
    "pop_origin_freqs = pop_origin_counts.div(row_totals, axis = 0)\n",
    "# Round to 2 significant figures\n",
    "pop_origin_freqs = signif(pop_origin_freqs, 2)\n",
    "print('Counts of origins:')\n",
    "display(pop_origin_counts)\n",
    "print('\\n\\nFrequencies of known origins:')\n",
    "display(pop_origin_freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a6e5a9a",
   "metadata": {},
   "source": [
    "### Diplotype clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb0482d1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd \n",
    "import plotly.express as px\n",
    "\n",
    "def _dipclust_concat_subplots(\n",
    "    figures,\n",
    "    width,\n",
    "    height,\n",
    "    row_heights,\n",
    "    title,\n",
    "    xaxis_range,\n",
    "):\n",
    "    from plotly.subplots import make_subplots  # type: ignore\n",
    "    import plotly.graph_objects as go  # type: ignore\n",
    "\n",
    "    # make subplots\n",
    "    fig = make_subplots(\n",
    "        rows=len(figures),\n",
    "        cols=1,\n",
    "        shared_xaxes=True,\n",
    "        vertical_spacing=0.02,\n",
    "        row_heights=row_heights,\n",
    "    )\n",
    "\n",
    "    for i, figure in enumerate(figures):\n",
    "        if isinstance(figure, go.Figure):\n",
    "            # This is a figure, access the traces within it.\n",
    "            for trace in range(len(figure[\"data\"])):\n",
    "                fig.append_trace(figure[\"data\"][trace], row=i + 1, col=1)\n",
    "        else:\n",
    "            # Assume this is a trace, add directly.\n",
    "            fig.append_trace(figure, row=i + 1, col=1)\n",
    "\n",
    "    fig.update_xaxes(visible=False)\n",
    "    fig.update_layout(\n",
    "        title=title,\n",
    "        width=width,\n",
    "        height=height,\n",
    "        hovermode=\"closest\",\n",
    "        plot_bgcolor=\"white\",\n",
    "        xaxis_range=xaxis_range,\n",
    "    )\n",
    "\n",
    "    return fig\n",
    "\n",
    "import numba\n",
    "from scipy.spatial.distance import squareform  # type: ignore\n",
    "\n",
    "@numba.njit(parallel=True)\n",
    "def multiallelic_diplotype_pdist(X, metric):\n",
    "    \"\"\"Optimised implementation of pairwise distance between diplotypes.\n",
    "\n",
    "    N.B., here we assume the array X provides diplotypes as genotype allele\n",
    "    counts, with axes in the order (n_samples, n_sites, n_alleles).\n",
    "\n",
    "    Computation will be faster if X is a contiguous (C order) array.\n",
    "\n",
    "    The metric argument is the function to compute distance for a pair of\n",
    "    diplotypes. This can be a numba jitted function.\n",
    "\n",
    "    \"\"\"\n",
    "    n_samples = X.shape[0]\n",
    "    n_pairs = (n_samples * (n_samples - 1)) // 2\n",
    "    out = np.zeros(n_pairs, dtype=np.float32)\n",
    "\n",
    "    # Loop over samples, first in pair.\n",
    "    for i in range(n_samples):\n",
    "        x = X[i, :, :]\n",
    "\n",
    "        # Loop over observations again, second in pair.\n",
    "        for j in numba.prange(i + 1, n_samples):\n",
    "            y = X[j, :, :]\n",
    "\n",
    "            # Compute distance for the current pair.\n",
    "            d = metric(x, y)\n",
    "\n",
    "            # Store result for the current pair.\n",
    "            k = square_to_condensed(i, j, n_samples)\n",
    "            out[k] = d\n",
    "\n",
    "    return out\n",
    "\n",
    "\n",
    "@numba.njit\n",
    "def square_to_condensed(i, j, n):\n",
    "    \"\"\"Convert distance matrix coordinates from square form (i, j) to condensed form.\"\"\"\n",
    "\n",
    "    assert i != j, \"no diagonal elements in condensed matrix\"\n",
    "    if i < j:\n",
    "        i, j = j, i\n",
    "    return n * j - j * (j + 1) // 2 + i - 1 - j\n",
    "\n",
    "\n",
    "@numba.njit\n",
    "def multiallelic_diplotype_mean_cityblock(x, y):\n",
    "    \"\"\"Compute the mean cityblock distance between two diplotypes x and y. The\n",
    "    diplotype vectors are expected as genotype allele counts, i.e., x and y\n",
    "    should have the same shape (n_sites, n_alleles).\n",
    "\n",
    "    N.B., here we compute the mean value of the distance over sites where\n",
    "    both individuals have a called genotype. This avoids computing distance\n",
    "    at missing sites.\n",
    "\n",
    "    \"\"\"\n",
    "    n_sites = x.shape[0]\n",
    "    n_alleles = x.shape[1]\n",
    "    distance = np.float32(0)\n",
    "    n_sites_called = np.float32(0)\n",
    "\n",
    "    # Loop over sites.\n",
    "    for i in range(n_sites):\n",
    "        x_is_called = False\n",
    "        y_is_called = False\n",
    "        d = np.float32(0)\n",
    "\n",
    "        # Loop over alleles.\n",
    "        for j in range(n_alleles):\n",
    "            # Access allele counts.\n",
    "            xc = np.float32(x[i, j])\n",
    "            yc = np.float32(y[i, j])\n",
    "\n",
    "            # Check if any alleles observed.\n",
    "            x_is_called = x_is_called or (xc > 0)\n",
    "            y_is_called = y_is_called or (yc > 0)\n",
    "\n",
    "            # Compute cityblock distance (absolute difference).\n",
    "            d += np.fabs(xc - yc)\n",
    "\n",
    "        # Accumulate distance for the current pair, but only if both samples\n",
    "        # have a called genotype.\n",
    "        if x_is_called and y_is_called:\n",
    "            distance += d\n",
    "            n_sites_called += np.float32(1)\n",
    "\n",
    "    # Compute the mean distance over sites with called genotypes.\n",
    "    if n_sites_called > 0:\n",
    "        mean_distance = distance / n_sites_called\n",
    "    else:\n",
    "        mean_distance = np.nan\n",
    "\n",
    "    return mean_distance\n",
    "\n",
    "def plot_dendrogram(\n",
    "    dist,\n",
    "    linkage_method,\n",
    "    count_sort,\n",
    "    distance_sort,\n",
    "    render_mode,\n",
    "    width,\n",
    "    height,\n",
    "    title,\n",
    "    line_width,\n",
    "    line_color,\n",
    "    marker_size,\n",
    "    leaf_data,\n",
    "    leaf_hover_name,\n",
    "    leaf_hover_data,\n",
    "    leaf_color,\n",
    "    leaf_symbol,\n",
    "    leaf_y,\n",
    "    leaf_color_discrete_map,\n",
    "    leaf_category_orders,\n",
    "    template,\n",
    "    y_axis_title,\n",
    "    y_axis_buffer,\n",
    "):\n",
    "    # Hierarchical clustering.\n",
    "    Z = sch.linkage(dist, method=linkage_method)\n",
    "\n",
    "    # Compute the dendrogram but don't plot it.\n",
    "    dend = sch.dendrogram(\n",
    "        Z,\n",
    "        count_sort=count_sort,\n",
    "        distance_sort=distance_sort,\n",
    "        no_plot=True,\n",
    "    )\n",
    "\n",
    "    # Compile the line coordinates into a single dataframe.\n",
    "    icoord = dend[\"icoord\"]\n",
    "    dcoord = dend[\"dcoord\"]\n",
    "    line_segments_x = []\n",
    "    line_segments_y = []\n",
    "    for ik, dk in zip(icoord, dcoord):\n",
    "        # Adding None here breaks up the lines.\n",
    "        line_segments_x += ik + [None]\n",
    "        line_segments_y += dk + [None]\n",
    "    df_line_segments = pd.DataFrame({\"x\": line_segments_x, \"y\": line_segments_y})\n",
    "\n",
    "    # Convert X coordinates to haplotype indices (scipy multiplies coordinates by 10).\n",
    "    df_line_segments[\"x\"] = (df_line_segments[\"x\"] - 5) / 10\n",
    "\n",
    "    # Plot the lines.\n",
    "    fig = px.line(\n",
    "        df_line_segments,\n",
    "        x=\"x\",\n",
    "        y=\"y\",\n",
    "        render_mode=render_mode,\n",
    "        template=template,\n",
    "    )\n",
    "\n",
    "    # Reorder leaf data to align with dendrogram.\n",
    "    leaves = dend[\"leaves\"]\n",
    "    n_leaves = len(leaves)\n",
    "    leaf_data = leaf_data.iloc[leaves]\n",
    "\n",
    "    # Add scatter plot to draw the leaves.\n",
    "    fig.add_traces(\n",
    "        list(\n",
    "            px.scatter(\n",
    "                data_frame=leaf_data,\n",
    "                x=np.arange(n_leaves),\n",
    "                y=np.repeat(leaf_y, n_leaves),\n",
    "                color=leaf_color,\n",
    "                symbol=leaf_symbol,\n",
    "                render_mode=render_mode,\n",
    "                hover_name=leaf_hover_name,\n",
    "                hover_data=leaf_hover_data,\n",
    "                template=template,\n",
    "                color_discrete_map=leaf_color_discrete_map,\n",
    "                category_orders=leaf_category_orders,\n",
    "            ).select_traces()\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Style the lines and markers.\n",
    "    line_props = dict(\n",
    "        width=line_width,\n",
    "        color=line_color,\n",
    "    )\n",
    "    marker_props = dict(\n",
    "        size=marker_size,\n",
    "    )\n",
    "    fig.update_traces(line=line_props, marker=marker_props)\n",
    "\n",
    "    # Style the figure.\n",
    "    fig.update_layout(\n",
    "        width=width,\n",
    "        height=height,\n",
    "        title=title,\n",
    "        autosize=True,\n",
    "        hovermode=\"closest\",\n",
    "        # I cannot get the xaxis title to appear below the plot, and when\n",
    "        # it's above the plot it often overlaps the title, so hiding it\n",
    "        # for now.\n",
    "        xaxis_title=None,\n",
    "        yaxis_title=y_axis_title,\n",
    "        showlegend=True,\n",
    "    )\n",
    "\n",
    "    # Style axes.\n",
    "    fig.update_xaxes(\n",
    "        mirror=False,\n",
    "        showgrid=False,\n",
    "        showline=False,\n",
    "        showticklabels=False,\n",
    "        ticks=\"\",\n",
    "        range=(-2, n_leaves + 2),\n",
    "    )\n",
    "    fig.update_yaxes(\n",
    "        mirror=False,\n",
    "        showgrid=False,\n",
    "        showline=False,\n",
    "        showticklabels=True,\n",
    "        ticks=\"outside\",\n",
    "        range=(leaf_y - y_axis_buffer, np.max(dcoord) + y_axis_buffer),\n",
    "    )\n",
    "\n",
    "    return fig, leaf_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f53cf42",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "df_samples = pd.read_csv(metadata_path, sep=\"\\t\")\n",
    "df_kdr = pd.read_csv(f'{wkdir}/results/kdr-origins/kdr_origins.csv', sep=\"\\t\", index_col=0)\n",
    "df_kdr = df_kdr.reset_index().rename(columns={'index': 'sample_id'})\n",
    "\n",
    "vcf_path = f\"{wkdir}/results/vcfs/amplicons/{dataset}.annot.vcf\"\n",
    "geno, pos, contig, df_samples, ref, alt = amp.load_vcf(vcf_path, metadata=df_samples)\n",
    "\n",
    "import json \n",
    "with open(f\"{wkdir}/config/metadata_colours.json\", 'r') as f:\n",
    "    color_mapping = json.load(f)\n",
    "\n",
    "bed_path = f\"{wkdir}/config/ag-vampir.bed\"\n",
    "df_bed = pd.read_csv(bed_path, sep=\"\\t\", header=None, names=['contig', 'start', 'end', 'amplicon_id', 'mutation', 'ref', 'alt'])\n",
    "\n",
    "# subset to VGSC SNPs\n",
    "vgsc_mask = df_bed.eval('mutation.str.contains(\"Vgsc\")').to_numpy()\n",
    "df_vgsc = df_bed[vgsc_mask]\n",
    "vgsc_start = df_vgsc['start'].min() - 200 \n",
    "vgsc_end = df_vgsc['end'].max() + 200\n",
    "vgsc_mask = np.logical_and(contig == '2L', np.logical_and(pos >= vgsc_start, pos <= vgsc_end))\n",
    "geno_vgsc = geno.compress(vgsc_mask, axis=0)\n",
    "pos_vgsc = pos[vgsc_mask]\n",
    "\n",
    "# remove invariant sites \n",
    "ac = geno_vgsc.count_alleles()\n",
    "is_seg = ac.is_segregating()\n",
    "geno_vgsc = geno_vgsc.compress(is_seg, axis=0)\n",
    "\n",
    "# remove highly missing sites \n",
    "missing_mask = geno_vgsc.is_missing().mean(axis=1) > 0.1\n",
    "missing_mask.sum()\n",
    "geno_vgsc = geno_vgsc.compress(~missing_mask, axis=0)\n",
    "\n",
    "# distances \n",
    "from scipy.spatial.distance import squareform\n",
    "import scipy.cluster.hierarchy as sch\n",
    "\n",
    "ac = allel.GenotypeArray(geno_vgsc).to_allele_counts(max_allele=3)\n",
    "X = np.ascontiguousarray(np.swapaxes(ac.values, 0, 1))\n",
    "dists = multiallelic_diplotype_pdist(X, metric=multiallelic_diplotype_mean_cityblock)\n",
    "dist_matrix = squareform(dists)\n",
    "df_dists = pd.DataFrame(dist_matrix, index=df_samples.index, columns=df_samples.index)\n",
    "na_mask = df_dists.isna().any()\n",
    "df_dists = df_dists.loc[~na_mask, ~na_mask]\n",
    "\n",
    "df_samples = df_samples.reset_index().merge(df_kdr[['sample_id', 'kdr_origin']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e9d2cb",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import scipy\n",
    "\n",
    "distance_metric = 'cityblock'\n",
    "leaf_color = 'location'\n",
    "\n",
    "fig_dendro, leaf_data = plot_dendrogram(\n",
    "    dist=scipy.spatial.distance.squareform(df_dists.values),\n",
    "    linkage_method=\"complete\",\n",
    "    count_sort=True,\n",
    "    distance_sort=False,\n",
    "    render_mode=\"svg\",\n",
    "    width=800,\n",
    "    height=500,\n",
    "    title=f\"{dataset} | Vgsc diplotype clustering\",\n",
    "    line_width=0.6,\n",
    "    line_color='black',\n",
    "    marker_size=5,\n",
    "    leaf_data=df_samples[~na_mask.to_numpy()],\n",
    "    leaf_hover_name=\"sample_id\",\n",
    "    leaf_hover_data=cohort_cols + ['kdr_origin'],\n",
    "    leaf_color=leaf_color,\n",
    "    leaf_symbol=None,\n",
    "    leaf_y=-0.01,\n",
    "    leaf_color_discrete_map=color_mapping[leaf_color],\n",
    "    leaf_category_orders=None,\n",
    "    template=\"simple_white\",\n",
    "    y_axis_title=f\"Distance ({distance_metric})\",\n",
    "    y_axis_buffer=0.1,\n",
    ")\n",
    "\n",
    "\n",
    "df_snps = pd.read_excel(f\"{wkdir}/results/vcfs/targets/ag-vampir-002-snps.xlsx\")\n",
    "df_vgsc_snps = df_snps.query(\"CHROM == '2L' and POS >= @vgsc_start and POS <= @vgsc_end\")\n",
    "df_vgsc_snps = df_bed.rename(columns={'end':'POS', 'contig':'CHROM'})[['CHROM', 'POS', 'mutation']].merge(df_vgsc_snps, on=['CHROM', 'POS'])\n",
    "df_vgsc_snps = df_vgsc_snps.set_index('mutation').iloc[:, 6:]\n",
    "df_vgsc_snps = df_vgsc_snps.query(\"~mutation.str.contains('AIM')\").query(\"~mutation.str.contains('tag')\")\n",
    "df_vgsc_snps = df_vgsc_snps.loc[:, leaf_data.sample_id.to_list()]\n",
    "df_vgsc_snps.index = df_vgsc_snps.index.str.replace(\"Vgsc_\", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affe0b60",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "figures = [fig_dendro]\n",
    "subplot_heights = [300]\n",
    "snp_row_height  = 20\n",
    "width = 800\n",
    "\n",
    "# kdr_fig = px.scatter(\n",
    "#     data_frame=kdr_leaf_data,\n",
    "#     x=np.arange(kdr_leaf_data.shape[0]),\n",
    "#     y=np.repeat(0, kdr_leaf_data.shape[0]),\n",
    "#     color='kdr_origin',\n",
    "#     hover_name='sample_id',\n",
    "#     symbol_sequence=['square'],\n",
    "#     # hover_data=leaf_hover_data,\n",
    "#     template='simple_white',\n",
    "#     # color_discrete_map=leaf_color_discrete_map,\n",
    "#     # category_orders=leaf_category_orders,\n",
    "# )\n",
    "\n",
    "# for f in kdr_fig.data:\n",
    "#     f.legendgroup = 'kdr_origin'\n",
    "\n",
    "# figures.append(kdr_fig)\n",
    "# subplot_heights.append(20)\n",
    "\n",
    "# het bar\n",
    "df_het = pd.DataFrame(\n",
    "    {\"sample_id\": samples, \"Sample Heterozygosity\": geno_vgsc.is_het().sum(axis=0) / geno_vgsc.is_called().sum(axis=0)}\n",
    ").set_index(\"sample_id\")\n",
    "\n",
    "# order according to dendrogram and transpose\n",
    "df_het = df_het.loc[leaf_data.sample_id, :].T\n",
    "het_trace = go.Heatmap(\n",
    "    z=df_het,\n",
    "    y=[\"Heterozygosity\"],\n",
    "    colorscale=\"Greys\",\n",
    "    showlegend=False,\n",
    "    showscale=False,\n",
    ")\n",
    "\n",
    "figures.append(het_trace)\n",
    "subplot_heights.append(30)\n",
    "\n",
    "snp_trace = go.Heatmap(\n",
    "    z=df_vgsc_snps[::-1].values,\n",
    "    y=df_vgsc_snps[::-1].index.to_list(),\n",
    "    colorscale=\"Greys\",\n",
    "    showlegend=False,\n",
    "    showscale=False,\n",
    ")\n",
    "\n",
    "figures.append(snp_trace)\n",
    "subplot_heights.append(snp_row_height * df_vgsc_snps.shape[0])\n",
    "\n",
    "height = sum(subplot_heights) + 50\n",
    "fig = _dipclust_concat_subplots(\n",
    "    figures=figures,\n",
    "    width=width,\n",
    "    height=height,\n",
    "    row_heights=subplot_heights,\n",
    "    title=f\"{dataset} | Vgsc diplotype clustering\",\n",
    "    xaxis_range=(0, df_dists.shape[0]),\n",
    ")\n",
    "\n",
    "fig[\"layout\"][\"yaxis\"][\"title\"] = f\"Distance ({distance_metric})\"\n",
    "\n",
    "aa_idx = len(figures)\n",
    "fig.add_hline(y=-0.5, line_width=1, line_color=\"grey\", row=aa_idx, col=1)\n",
    "for i, y in enumerate(df_vgsc_snps.index.to_list()):\n",
    "    fig.add_hline(y=i+0.5, line_width=1, line_color=\"grey\", row=aa_idx, col=1)\n",
    "\n",
    "fig['layout'][f'yaxis{aa_idx}']['title']=f'Vgsc mutations'\n",
    "fig.update_xaxes(showline = True, linecolor = 'grey', linewidth = 1, row = aa_idx, col = 1, mirror = True)\n",
    "fig.update_yaxes(showline = True, linecolor = 'grey', linewidth = 1, row = aa_idx, col = 1, mirror = True)\n",
    "#fig.write_image(f\"{wkdir}/results/kdr_dipclust.png\", scale=2)\n",
    "fig.show()"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
